src/useObservableProps.ts

```diff
@@ -0,0 +1,187 @@
+import {
+    action,
+    isObservable,
+    isObservableArray,
+    isObservableMap,
+    isObservableObject,
+    observable,
+    remove,
+    set
+} from "mobx"
+import { useMemo } from "react"
+
+type LocalObservables = WeakMap<any, boolean>
+
+const EMPTY_OBJECT = {}
+
+export type UseObservablePropsMode<P> =
+    | "shallow" // the reference is not changed and the props (primitives, objects, maps and arrays) are turned into a shallowly observable object
+    | "deep" // the reference is not changed and props (primitives, objects, maps and arrays) are turned into a deeply observable object
+    | {
+          deepProps: Array<keyof P> // like 'shallow', except some properties are turned into deep observables 'opt-in'
+      }
+
+export function useObservableProps<P>(props: P, mode: UseObservablePropsMode<P>): P {
+    const obsProps = useMemo(() => {
+        return toObservablePropsAction(props, mode)
+    }, [])
+
+    obsProps.update(props)
```

Comment:
It will do the setup on mount within `useMemo` and then do the update immediately knowing that nothing is changed at that point?

Reply:
done